File: R/cassandra_table.R
#' Cassandra Table
#'
#' Access scala handles to cassandra tables
#' Using the spark-cassandra-connector
#'
#' @rdname CassandraTable
#' @include
NULL

#' Get a Cassandra table
#'
#' @export
cql_get_table <- function(sc, keyspace, table, select_cols = list()) {

  sparklyr::invoke_static(sc, "CQLConnect.CassandraTable", "get_table", sc$spark_context, keyspace, table, select_cols)

}

#' Test joinWithCassandraTable
#'
#' @export
cql_get_obc_model <- function(sc, keyspace, line, date, select_cols = list()) {

  cols_reserved = c("line", "vehicle_id_command", "date", "vcc", "channel")
  reserved_cols_used = cols_reserved %in% select_cols

  # However the date is passed in, the scala method expects a character
  date = as.character(date)
  line = as.integer(line)

  if(any(reserved_cols_used))
    stop("Partition keys not necessary to specify, keys used: ", paste0(cols_reserved[reserved_cols_used], collapse = ", "))

  if(length(select_cols) > 0)
    select_cols = c(as.list(cols_reserved), select_cols)

  sparklyr::invoke_static(sc, "CQLConnect.CassandraTable", "get_obc_model", sc$spark_context, keyspace, line, date, select_cols)

}

#' Get a Cassandra table
#'
#' @export
cql_get_schema <- function(sc, keyspace, table, select_cols = list()) {

  sparklyr::invoke_static(sc, "CQLConnect.CassandraTable", "get_schema", sc$spark_context, keyspace, table)

}

#' Test joinWithRTable
#'
#' @export
cql_joinWithRTable <- function(sc, tbl_partitions, keyspace, table, select_cols = list()) {

  dataset = sparklyr::spark_dataframe(tbl_partitions)

  writer = sparklyr::invoke_static(sc, "CQLConnect.CassandraTable", "getWriter")
  reader = sparklyr::invoke_static(sc, "CQLConnect.CassandraTable", "getReader")

  sparklyr::invoke_static(sc, "CQLConnect.CassandraTable", "joinWithRTable", 
    sc$spark_context, dataset, keyspace, table, select_cols, writer, reader)

}

File: java/CassandraTable.scala
package CQLConnect

import java.sql
import java.util.Date
import scala.reflect.ClassTag

import com.datastax.spark.connector._
import com.datastax.spark.connector.cql._

import org.apache.spark.sql.cassandra._
import org.apache.spark.sql.types._
import org.apache.spark.sql.{Dataset, Row, SparkSession}
import org.apache.spark.SparkContext

import org.apache.spark.rdd.RDD

import CQLConnect.DateUtils.DateFormatter

object CassandraTable {

  // Map Cassandra elements to Spark
  def convertToSpark(element:Any): Any = element match {
    case time: org.joda.time.LocalDate => new sql.Date(time.toDateTimeAtStartOfDay().getMillis) // Convert to java.sql.Date
    case date: java.util.Date => new sql.Timestamp(date.getTime)
    case uuid: java.util.UUID => uuid.toString()
    case other => other
  }

  // Get a Cassandra table schema
  def get_schema(sc: SparkContext, keyspace: String, table: String) = {
    val spark =  SparkSession.builder().getOrCreate()
    spark.read.cassandraFormat(table, keyspace).load.schema
  }

  def get_table(sc: SparkContext, keyspace: String, table: String, select_cols: Array[String]) = {
    val spark =  SparkSession.builder().getOrCreate()
    import spark.implicits._
    // Cassandra schema and table
    var schema = get_schema(sc, keyspace, table)
    var cass_table = sc.cassandraTable(keyspace, table)
    // Select columns is necessary
    if (select_cols.length > 0) {
      schema = StructType(schema.filter(x => select_cols.contains(x.name)))
      cass_table = cass_table.select(select_cols.map(ColumnName(_)):_*)
    }
    // Convert Cassandra values to Spark and return
    val spk_cass = cass_table.map{ case cassandraRow => Row(cassandraRow.columnValues.map(convertToSpark):_*) }
    spark.createDataFrame(spk_cass, schema)
  }

  // Hardcode function to pull obc_model
  //  - Read in partitions model from C*
  //  - Map into a table with an added date
  //  - Repartition by obc_model partitioner to distribute load correctly
  //  - Join with C8 table to return the subset RDD
  //  - Filter if necessary
  //  - Map C* values to Spark Dataset to hand to Sparklyr
  def get_obc_model ( sc: SparkContext, keyspace: String, int_line: Int, str_date: String, select_cols: Array[String]) = {
    val spark =  SparkSession.builder().getOrCreate()
    import spark.implicits._

    var date: sql.Date = new sql.Date(DateFormatter.parse(str_date).getTime())
    var schema = get_schema(sc, keyspace, "obc_model")

    // Repartitioning to the spark cluster as
    // partitions_obc_model is only distributed across 2 nodes
    var cass_join =
      sc.cassandraTable(keyspace, "partitions_obc_model")
        .where("line = ?", int_line)
        // TODO: Case class this
        .map{ case cassandraRow => (
          cassandraRow.getInt("line"),
          cassandraRow.getInt("vehicle_id_command"),
          date,
          cassandraRow.getInt("vcc"),
          cassandraRow.getInt("channel")
        )}
        .repartitionByCassandraReplica(keyspace, "obc_model")
        .joinWithCassandraTable(keyspace, "obc_model")

    if (select_cols.length > 0) {
      schema = StructType(schema.filter(x => select_cols.contains(x.name)))
      cass_join = cass_join.select(select_cols.map(ColumnName(_)):_*)
    }

    val spk_cass_join = cass_join.map{ case(_, cassandraRow) => Row(cassandraRow.columnValues.map(convertToSpark):_*)}
    spark.createDataFrame(spk_cass_join, schema)
  }

  // Can the partitions table be supplied via Sparklyr?
  def joinWithRTable[T <: Serializable: ClassTag] (
    sc: SparkContext,
    dataset: Dataset[T],
    keyspace: String,
    table: String,
    select_cols: Array[String])(
    implicit
      rwf: writer.RowWriterFactory[T],
      rrf: rdd.reader.RowReaderFactory[T]
    ) = {

    val spark =  SparkSession.builder().getOrCreate()
    import spark.implicits._

    val my_rdd = dataset.rdd
    var schema = get_schema(sc, keyspace, table)
    var cass_join =
      my_rdd
        .repartitionByCassandraReplica(keyspace, table)
        .joinWithCassandraTable(keyspace, table)

    if (select_cols.length > 0) {
      schema = StructType(schema.filter(x => select_cols.contains(x.name)))
      cass_join = cass_join.select(select_cols.map(ColumnName(_)):_*)
    }

    cass_join
  }

}
